browser/ipfs/content_browser_client_helper_unittest.cc

```diff
@@ -0,0 +1,230 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ipfs/content_browser_client_helper.h"
+
+#include <memory>
+#include <vector>
+
+#include "base/test/bind_test_util.h"
+#include "base/test/scoped_feature_list.h"
+#include "brave/browser/profiles/brave_profile_manager.h"
+#include "brave/browser/profiles/brave_unittest_profile_manager.h"
+#include "brave/browser/profiles/profile_util.h"
+#include "brave/browser/tor/buildflags.h"
+#include "brave/common/pref_names.h"
+#include "brave/components/ipfs/browser/features.h"
+#include "brave/components/ipfs/common/ipfs_constants.h"
+#include "chrome/browser/prefs/browser_prefs.h"
+#include "chrome/test/base/chrome_render_view_host_test_harness.h"
+#include "chrome/test/base/scoped_testing_local_state.h"
+#include "chrome/test/base/testing_browser_process.h"
+#include "chrome/test/base/testing_profile.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/test/browser_task_environment.h"
+#include "content/public/test/mock_navigation_handle.h"
+#include "content/public/test/test_utils.h"
+#include "content/public/test/web_contents_tester.h"
+#include "testing/gtest/include/gtest/gtest.h"
+#include "url/gurl.h"
+
+namespace {
+
+const GURL& GetIPFSURI() {
+  static const GURL ipfs_url(
+      "ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/wiki/Vincent_van_Gogh.html");  // NOLINT
+  return ipfs_url;
+}
+
+const GURL& GetIPFSGatewayURL() {
+  static const GURL ipfs_url(
+      "https://dweb.link/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/wiki/Vincent_van_Gogh.html");  // NOLINT
+  return ipfs_url;
+}
+
+const GURL& GetIPFSLocalURL() {
+  static const GURL ipfs_url(
+      "http://127.0.0.1:8080/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/wiki/Vincent_van_Gogh.html");  // NOLINT
```

Comment:
There is a small nuance regarding `localhost` URLs, thought its worth mentioning here:

- `127.0.0.1` is a path-based gateway
  -  `http://127.0.0.1:8080/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/wiki/Vincent_van_Gogh.html` will return HTTP 200 with HTML response
- `localhost` provides origin isolation via subdomain gateway
  - `http://localhost:8080/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/wiki/Vincent_van_Gogh.html` will return HTTP 301 with redirect to unique Origin at localhost subdomain:  `http://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq.ipfs.localhost:8080/wiki/Vincent_van_Gogh.html`

Due to this, adding a separate test for `*.ipfs.localhost:8080`  to document the behavior of both types of local gateways would be useful.

Reply:
Yep understood thanks @lidel.  Currently we will only show `ipfs://` when the user types in `ipfs://` (i.e. we won't reverse resolve to ipfs:// if the user puts in localhost or 127.0.0.1.  Also currently we always resolve ipfs:// internally to 127.0.0.1:8080.  So the test is currently consistent with that.  But I'll keep this in mind for future changes.

Reply:
SGTM. Showing `ipfs://` only if it was explicitly used is a good call :+1: